src/pyspiffe/spiffe_id/spiffe_id.py

```diff
@@ -1,17 +1,167 @@
+            >>> print(spiffe_id_2)
+            spiffe://example.org/path1/path2/element
+        """
+
+        if trust_domain is None:
+            raise ValueError('SPIFFE ID: trust domain cannot be empty.')
+
+        if not isinstance(trust_domain, TrustDomain):
+            raise ValueError(
+                'SPIFFE ID: trust_domain argument must be a TrustDomain instance.'
+            )
+
+        result = SpiffeId()
+        result.__set_trust_domain(trust_domain)
+
+        if path_segments is not None:
+            result.__set_path(path_segments)
+
+        cls.parse_and_validate_uri(str(result))
+        return result
+
+    def __eq__(self, other: Any) -> bool:
+        if not isinstance(other, self.__class__):
+            return False
+        return (
+            self.__trust_domain == other.__trust_domain and self.__path == other.path()
+        )
+
+    def __str__(self) -> str:
+        if self.__path:
+            return '{}://{}{}'.format(
+                SPIFFE_SCHEME, self.__trust_domain.name(), self.__path
+            )
+        return '{}://{}'.format(SPIFFE_SCHEME, self.__trust_domain.name())
+
+    # path_segments can be an array of path segments or a single string representing a path
+    def __set_path(self, path: Any) -> None:
+        if isinstance(path, list):
+            self.__path = ''
+            for s in path:
+                self.__path += self.normalize_path(s)
+            return
+        self.__path = self.normalize_path(path)
+
+    def __set_trust_domain(self, trust_domain: TrustDomain) -> None:
+        self.__trust_domain = trust_domain
+
+    def path(self) -> str:
+        return self.__path
+
+    def trust_domain(self) -> TrustDomain:
+        return self.__trust_domain
+
+    def is_member_of(self, trust_domain: TrustDomain) -> bool:
+        return self.__trust_domain == trust_domain
+
+    @classmethod
+    def parse_and_validate_uri(cls, spiffe_id: str) -> dict:
+        if len(spiffe_id) > SPIFFE_ID_MAXIMUM_LENGTH:
+            raise ValueError(
+                'SPIFFE ID: maximum length is {} bytes.'.format(
+                    SPIFFE_ID_MAXIMUM_LENGTH
+                )
+            )
+
+        uri = parse(spiffe_id.strip(), rule='URI')
+
+        scheme = uri.get('scheme')
+        if scheme.lower() != SPIFFE_SCHEME:
+            raise ValueError('SPIFFE ID: invalid scheme: expected spiffe.')
+
+        query = uri.get('query')
+        if query:
+            raise ValueError('SPIFFE ID: query is not allowed.')
+
+        fragment = uri.get('fragment')
+        if fragment:
+            raise ValueError('SPIFFE ID: fragment is not allowed.')
+
+        authority = uri.get('authority')
+        if not authority:
+            raise ValueError('SPIFFE ID: trust domain cannot be empty.')
+
+        # has user@info:
+        if '@' in authority:
+            raise ValueError('SPIFFE ID: user info is not allowed.')
```

Comment:
Should info be spelled complete instead: information?

Reply:
amended to `userinfo`, that's the component name of the URI in the RFC 3986.

Reply:
Thanks Max, much better!